src/main/java/org/checkerframework/checker/builder/TypesafeBuilderAnnotatedTypeFactory.java

```diff
@@ -119,6 +126,100 @@ public Void visitMethodInvocation(
+        return super.visitExecutable(t, p);
+      }
+
+      boolean inAutoValueBuilder = hasAnnotation(enclosingClass, AutoValue.Builder.class);
+
+      if (inAutoValueBuilder) {
+        // get the name of the method
+        String methodName = methodTree.getName().toString();
+
+        ClassTree autoValueClass =
+            TreeUtils.enclosingClass(getPath(enclosingClass).getParentPath());
+
+        assert hasAnnotation(autoValueClass, AutoValue.class)
+            : "class " + autoValueClass.getSimpleName() + " is missing @AutoValue annotation";
+
+        if ("build".equals(methodName)) {
+          // determine the required properties and add a corresponding @CalledMethods annotation
+          List<String> requiredProperties = getRequiredProperties(autoValueClass);
+          AnnotationMirror newCalledMethodsAnno =
+              createCalledMethodsForProperties(requiredProperties);
+          t.getReceiverType().addAnnotation(newCalledMethodsAnno);
+        }
+      }
+
+      return super.visitExecutable(t, p);
+    }
+
+    /**
+     * computes the required properties of an @AutoValue class
+     *
+     * @param autoValueClass
+     * @return
+     */
+    private List<String> getRequiredProperties(ClassTree autoValueClass) {
+      List<String> requiredPropertyNames = new ArrayList<>();
+      for (Tree member : autoValueClass.getMembers()) {
+        if (member.getKind() == Tree.Kind.METHOD) {
+          MethodTree methodTree = (MethodTree) member;
+          // should be an instance method
+          if (!methodTree.getModifiers().getFlags().contains(Modifier.STATIC)) {
+            String name = methodTree.getName().toString();
+            if (!IGNORED_METHOD_NAMES.contains(name)
+                && !methodTree.getReturnType().toString().equals("void")) {
+              // shouldn't have a nullable return
+              List<? extends AnnotationTree> annotations =
+                  methodTree.getModifiers().getAnnotations();
+              boolean hasNullable =
+                  annotations.stream()
+                      .map(TreeUtils::annotationFromAnnotationTree)
+                      .anyMatch(anm -> AnnotationUtils.annotationName(anm).endsWith(".Nullable"));
+              if (!hasNullable) {
+                requiredPropertyNames.add(name);
+              }
+            }
+          }
+        }
+      }
+      return requiredPropertyNames;
+    }
+
+    /**
+     * creates a @CalledMethods annotation for the given property names, converting the names to the
+     * corresponding setter method name in the Builder
+     *
+     * @param propertyNames the property names
+     * @return the @CalledMethods annotation
+     */
+    public AnnotationMirror createCalledMethodsForProperties(final List<String> propertyNames) {
+      String[] calledMethodNames =
+          propertyNames.stream()
+              .map((prop) -> "set" + prop.substring(0, 1).toUpperCase() + prop.substring(1))
```

Comment:
Does AutoValue permit the user to change the names of generated methods? I know Lombok does, and it's a huge pain to deal with.

If so, I'm not sure the assumptions about the names of methods here will hold. I'm not sure we can really do any better, though, so maybe it's okay.

Reply:
I am actually not 100% sure what AutoValue allows in builders.  I'd have to dig into the implementation to check.  But in my experience I've only seen the `setXXX()` syntax.  We will probably have to handle a bunch of additional corner cases when we test on real code, so I'm not going to pre-emptively investigate.

Reply:
That seems reasonable.